README.md
# Scroll Speed Limit

CHANGELOG.md
# CHANGELOG

## [1.0.1] - 2021-04-07
- avoid "offsetParent"

## [1.0.0] - 2021-04-07
- start

test/src/main.ts
import setScrollMaxSpeed from "../../src/index";

(window as any).scrollSpeedLimit = setScrollMaxSpeed;
setScrollMaxSpeed(0.5);

// const el = document.body.firstElementChild!;
// let lastOffsetTop = el.getBoundingClientRect().top;
// let startTransition: number = 0;
// let startTransitionOffset: number | null = null;
// let lastEqual = true;
// let deltasY: number[] = [];
// requestAnimationFrame(function test() {
//     const offsetTop = el.getBoundingClientRect().top;
//     if (lastOffsetTop === offsetTop) {
//         if (!lastEqual) {
//             const data = {
//                 start: {
//                     offsetTop: startTransitionOffset!,
//                     time: startTransition
//                 },
//                 end: {
//                     offsetTop,
//                     time: Date.now()
//                 }
//             };
//             console.log(
//                 data.end.offsetTop - data.start.offsetTop,
//                 data.end.time - data.start.time,
//                 `${
//                     (data.end.offsetTop - data.start.offsetTop) /
//                     ((data.end.time - data.start.time) * 0.001)
//                 }px/s`,
//                 deltasY
//                 // deltasY.reduce((prev, current) => current + prev, 0) / deltasY.length
//             );
//             startTransitionOffset = null;
//         }
//         lastEqual = true;
//     } else {
//         if (startTransitionOffset == null) {
//             console.log("starting transition");
//             startTransitionOffset = lastOffsetTop;
//             startTransition = Date.now();
//             deltasY = [ (window as any).lastDeltaY ];
//         } else {
//             if ((window as any).lastDeltaY != null) {
//                 deltasY.push((window as any).lastDeltaY);
//             }
//         }
//         lastEqual = false;
//     }
//     lastOffsetTop = offsetTop;
//     requestAnimationFrame(test);
// });

test/scripts/main.js
(function (factory) {
    typeof define === 'function' && define.amd ? define(factory) :
    factory();
}((function () { 'use strict';

    var maxSpeed = null;
    var supportsPassive = (function () {
        var supportsPassive = false;
        try {
            window.addEventListener("test", null, Object.defineProperty({}, "passive", {
                get: function () { supportsPassive = true; }
            }));
        }
        catch (e) { }
        return supportsPassive;
    })();
    var wheelOpt = supportsPassive ? { passive: false } : false;
    var wheelEvent = "onwheel" in document ? "wheel" : "mousewheel";
    function easeInOutSine(x) {
        return -(Math.cos(Math.PI * x) - 1) / 2;
    }
    var startScrollTime = null;
    var targetScrollTime = null;
    var startScroll = null;
    var targetScroll = null;
    function animationFrame() {
        if (targetScroll == null) {
            return;
        }
        var perc = (Date.now() - startScrollTime) / (targetScrollTime - startScrollTime);
        if (perc >= 1) {
            window.scrollTo(0, targetScroll);
            startScrollTime = null;
            targetScrollTime = null;
            startScroll = null;
            targetScroll = null;
            return;
        }
        window.scrollTo(0, ((targetScroll - startScroll) * easeInOutSine(perc)) + startScroll);
        requestAnimationFrame(animationFrame);
    }
    function customScroll(event) {
        event.preventDefault();
        var deltaY = event.deltaY * (event.deltaMode === 1 ? 40 : 1);
        var scrollParent = document.scrollingElement || document.firstElementChild || document.body;
        if (startScroll == null) {
            startScroll = scrollParent.scrollTop;
            startScrollTime = Date.now();
            targetScroll = startScroll;
            targetScrollTime = startScrollTime;
        }
        targetScroll = targetScroll + deltaY;
        targetScrollTime = targetScrollTime + (Math.abs(deltaY) / maxSpeed);
        requestAnimationFrame(animationFrame);
    }
    var keys = {
        37: 0,
        38: -1,
        39: 0,
        40: 1,
        32: 1
    };
    function preventDefaultForScrollKeys(e) {
        if (Object.prototype.hasOwnProperty.call(keys, e.keyCode)) {
            var event = {
                deltaY: keys[e.keyCode] * (e.shiftKey ? -1 : 1),
                deltaMode: 1,
                preventDefault: e.preventDefault.bind(e),
                target: document.body
            };
            customScroll(event);
            return false;
        }
    }
    function addEventListener() {
        window.addEventListener(wheelEvent, customScroll, wheelOpt);
        window.addEventListener("keydown", preventDefaultForScrollKeys, false);
    }
    function removeEventListener() {
        window.removeEventListener(wheelEvent, customScroll, wheelOpt);
        window.removeEventListener("keydown", preventDefaultForScrollKeys, false);
    }
    function setScrollMaxSpeed(limit) {
        if (typeof limit !== "number" || isNaN(limit) || limit <= 0) {
            limit = null;
        }
        else {
            limit = Math.round(limit);
        }
        if (limit === maxSpeed) {
            return;
        }
        if (limit == null) {
            if (maxSpeed != null) {
                removeEventListener();
            }
        }
        else {
            if (maxSpeed == null) {
                addEventListener();
            }
        }
        maxSpeed = limit;
    }

    window.scrollSpeedLimit = setScrollMaxSpeed;
    setScrollMaxSpeed(0.5);

})));


rollup.config.js
const commonjs = require('@rollup/plugin-commonjs');
const typescript = require('@rollup/plugin-typescript');
const nodeResolve = require("@rollup/plugin-node-resolve").default;

export default [
    {
        input: "src/index.ts",
        plugins: [
            nodeResolve(),
            typescript(),
            commonjs()
        ],
        output: [
            {
                name: "scrollSpeedLimit",
                file: "dist/index.js",
                format: "umd"
            },
            {
                file: "dist/index.es.js",
                format: "es"
            }
        ]
    },
    {
        input: "test/src/main.ts",
        plugins: [
            nodeResolve(),
            commonjs(),
            typescript()
        ],
        output: {
            file: "test/scripts/main.js",
            format: "umd"
        }
    }
];

src/index.ts
let maxSpeed: number | null = null;

// modern Chrome requires { passive: false } when adding event
const supportsPassive = (() => {
    let supportsPassive = false;
    try {
        window.addEventListener("test", null as any, Object.defineProperty({}, "passive", {
            get: function () { supportsPassive = true; }
        }));
    } catch (e) { }
    return supportsPassive;
})();

const wheelOpt = supportsPassive ? { passive: false } : false;
const wheelEvent = "onwheel" in document ? "wheel" : "mousewheel";

function easeInOutSine(x: number): number {
    return -(Math.cos(Math.PI * x) - 1) / 2;
}

let startScrollTime: number | null = null;
let targetScrollTime: number | null = null;
let startScroll: number | null = null;
let targetScroll: number | null = null;
function animationFrame() {
    if (targetScroll == null) {
        return;
    }
    const perc = (Date.now() - startScrollTime!) / (targetScrollTime! - startScrollTime!);
    if (perc >= 1) {
        window.scrollTo(0, targetScroll);
        startScrollTime = null;
        targetScrollTime = null;
        startScroll = null;
        targetScroll = null;
        return;
    }
    window.scrollTo(
        0,
        ((targetScroll - startScroll!) * easeInOutSine(perc)) + startScroll!
    );
    requestAnimationFrame(animationFrame);
}
function customScroll(event) {
    event.preventDefault();
    
    const deltaY = event.deltaY * (event.deltaMode === 1 ? 40 : 1);
    const scrollParent = document.scrollingElement || document.firstElementChild || document.body;
    if (startScroll == null) {
        startScroll = scrollParent.scrollTop;
        startScrollTime = Date.now();
        targetScroll = startScroll;
        targetScrollTime = startScrollTime;
    }
    targetScroll = targetScroll! + deltaY;
    targetScrollTime = targetScrollTime! + (Math.abs(deltaY) / maxSpeed!);
    requestAnimationFrame(animationFrame);
}

// thanks to https://stackoverflow.com/a/4770179/9228492
const keys = {
    37: 0, // left
    38: -1, // up
    39: 0, // right
    40: 1, // down
    32: 1  // spacebar
};

function preventDefaultForScrollKeys(e) {
    if (Object.prototype.hasOwnProperty.call(keys, e.keyCode)) {
        var event = {
            deltaY: keys[e.keyCode] * (e.shiftKey ? -1 : 1),
            deltaMode: 1,
            preventDefault: e.preventDefault.bind(e),
            target: document.body
        };
        customScroll(event);
        return false;
    }
}

function addEventListener() {
    window.addEventListener(wheelEvent, customScroll, wheelOpt); // modern desktop
    window.addEventListener("keydown", preventDefaultForScrollKeys, false);
}
function removeEventListener() {
    window.removeEventListener(wheelEvent, customScroll, wheelOpt as any); // modern desktop
    window.removeEventListener("keydown", preventDefaultForScrollKeys, false);
}

export default function setScrollMaxSpeed(limit: number | null): void {
    if (typeof limit !== "number" || isNaN(limit) || limit <= 0) {
        limit = null;
    } else {
        limit = Math.round(limit);
    }
    if (limit === maxSpeed) {
        return;
    }
    if (limit == null) {
        if (maxSpeed != null) {
            removeEventListener();
        }
    } else {
        if (maxSpeed == null) {
            addEventListener();
        }
    }
    maxSpeed = limit;
}